### ISSUE TITLE
Load the Chrome extension for end-to-end tests

### ISSUE THREAD

**Issue by ProfJanetDavis** [glam-lab/degender-the-web#101]
Currently, `test/e2e/helper.html` explicitly includes `src/main.js` and calls `main()`. This works fine as long as the extension modifies only page content. 

However, the resolution of #68 will use more Chrome extension features. To test those features, the extension must be loaded for end-to-end testing.

I already tried to do this once, following some of the advice in [this StackOverflow answer](https://stackoverflow.com/questions/48089670/detect-and-test-chrome-extension-using-puppeteer). It didn't work the first time; I decided to move on since there was a working alternative. Let's try again now that there is a need.

[This tutorial](https://dev.to/gokatz/automate-the-ui-testing-of-your-chrome-extension-52e1) might be more helpful than the answer I was looking at before.

**Comment by ProfJanetDavis** [comment]
@irhawk, I'm not going to assign this to you, but you are welcome to assign yourself if you want to give it a try. I'm about to be traveling and I have a couple of deadlines coming up, so I'm not sure when I'll get to it.

Note that we are using `npm`, not `yarn`, for package management. Puppeteer is already installed, in any case.

**Comment by beszel** [comment]
Sure thing! As you said, #68 needs this for testing. I'll take care of this before continuing work on that.

**Comment by beszel** [comment]
The tutorial sample has a 2-second delay so that the extension can apply to the page. After making some tests fail, I've had to do the same. The only tests now failing are related to header buttons, and those will soon be moved to the popup. I don't think the popup has any of the same timing issues.
Overall, Puppeteer is no longer running headless! It loads the extension just like Chrome does normally. I'm going to wait to resolve this simultaneously with #68 so that I don't need to fix the header tests. If that doesn't sound right, I could disable the header tests to resolve this sooner instead.

**Comment by ProfJanetDavis** [comment]
Resolving this simultaneously with #68 sounds like a good plan.


I don't think the StackOverflow answer mentioned needing a delay - or if it did, I wasn't reading carefully enough. A couple thoughts on that arbitrary delay:
- It's not ideal to have a time-based delay. It's better to wait for some meaningful event to occur.  If you have obtained a stable extension id, you might be able to wait for `popup.html` to load, instead of waiting on a timer. (This is from that StackOverflow answer linked above. I wouldn't spend a lot of time on this, but once you've created `popup.html`, try it once and see if it works.)
<pre>const page = await browser.newPage();
await page.goto('chrome-extension://ipfiboohojhbonenbbppflmpfkakjhed/popup.html');</pre>
- Two seconds is a long time compared to how long it takes the tests themselves to run. You might also/instead try experimenting a little to see if a shorter delay would work. (One second? Half a second?)
- The delay should be documented in `DEVELOPMENT.md` as something to check if end-to-end tests are failing unexpectedly.

**Comment by ProfJanetDavis** [comment]
Also, I'm looking at the [Puppeteer API docs](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#working-with-chrome-extensions) and I noted this:
> NOTE It is not yet possible to test extension popups or content scripts.

**Comment by beszel** [comment]
The StackOverflow answer is addressing how to test an extension popup, not a content script; The delay is only for the content script to finish loading. The delay isn't needed when the content script is called from the test page itself. My first thought was that we could run two instances of Puppeteer — headless for the content script and full for the popup — but the content script needs to respond to user input in the popup. I also tried setting the content script to apply at `document_end` rather than `document_idle`, but that didn't help.
I haven't been able to find anything about an event that happens when the extension has loaded. I'll look a bit further before deciding to live with the delay.

**Comment by ProfJanetDavis** [comment]
That makes sense. Since the delay works, don’t put too much effort into exploring other possibilities - maybe another 15 minutes of research and another 15 minutes to make an issue to perhaps resolve in the future. 

--
Janet Davis

**Comment by beszel** [comment]
I've left the delay in for now, but a better solution might use [`page.waitForSelector`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pagewaitforselectorselector-options) to see if `div#text` has any replacements in it.

**Comment by ProfJanetDavis** [comment]
I like that idea! If we know there should be replacements, we can wait for them. For uniformity, we could add a DOM element not visible to the user to indicate the content script has run. Or find a way to wait until the extension status variable (see #68) has a value. 

--
Janet Davis

**Comment by beszel** [comment]
I've removed the delay while fixing another issue. There are a couple of popup tests and they're now passing.

**Comment by ProfJanetDavis** [comment]
Woohoo!

--
Janet Davis

**Comment by beszel** [comment]
@ProfJanetDavis I need to [copy the key to the manifest](https://developer.chrome.com/apps/app_identity#copy_key). Do you have the key? It looks like many other extensions leave the key in the manifest for release, although it could have been added automatically. It might only make the extension ID consistent when installing from source, which is what we want.

**Comment by beszel** [comment]
There's an odd inconvenience when running end-to-end tests with full Puppeteer: Puppeteer opens Chromium, but it quickly fails and leaves the browser open to a blank tab.
```
  1) "before all" hook in "{root}"
  2) "after all" hook in "{root}"

  0 passing (2s)
  2 failing

  1) "before all" hook in "{root}":
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
      at listOnTimeout (internal/timers.js:531:17)
      at processTimers (internal/timers.js:475:7)

  2) "after all" hook in "{root}":
     ReferenceError: browser is not defined
      at Context.<anonymous> (test/e2e/bootstrap.js:47:5)
      at processImmediate (internal/timers.js:439:21)
```
After getting this error, I can interrupt with `ctrl-C` and run the tests again. They pass on the second attempt, and I'm sure that all the right tests are running. After a few minutes without running tests, this pattern restarts.

I could use any ideas you have about this; I have no idea what causes it.

**Comment by ProfJanetDavis** [comment]
I've seen transient errors like this when my laptop is in power saving mode due to running low on battery charge. Is that a possible explanation here?

**Comment by ProfJanetDavis** [comment]
@irhawk there is no key. The extension is not using the [Chrome Identity API](https://developer.chrome.com/apps/identity) linked from [the documentation you linked above](https://developer.chrome.com/apps/app_identity#copy_key). 

This doesn't seem like a good direction to me, even if it does what you hope it would. We would need to disclose in the privacy policy if we're associating use of the extension with an authenticated Google Apps user.
